keps/sig-etcd/4578-server-feature-gate/README.md

```diff
@@ -0,0 +1,364 @@
+
+But if the bug fix is very critical, it could be discussed in a case-by-case manner.
+
+#### Could the lifecycle of a feature change in patch versions?
+
+Kubernetes have a minor release every 3 months, while the cadence of etcd minor releases is much less frequent. The question is: do we have to wait for years before graduating a new feature?
+
+We think we should still stick to the common practice of not changing the lifecycle of a feature in patch versions. Because:
+* changing the lifecycle of a feature is an API change. According to the [etcd Operations Guide](https://etcd.io/docs/v3.5/op-guide/versioning/), only new minor versions may add additional features to the API.
+* bugs in etcd could be hard to detect, and the reliability and robustness of etcd is more important than speed. A long history of testing through practical adoption a new feature is beneficial.
+
+With the feature gate in place, we could consider increasing the etcd release cadence because it would be easier to add new features and less risky to release new features.
+
+### Risks and Mitigations
+
+* As discussed before, we are deferring cluster level feature support to a follow-up KEP.
+
+## Design Details
+
+### New APIs
+
+A new `--feature-gates` command line argument would be added to start the etcd server, with format like `--feature-gates=featureA=true,featureB=false`. The flag can also be set in the `config-file`.
+
+New Prometheus gauge metrics would be added to monitor and query if a feature is enabled for the server. 
+For easier usage in Kubernetes, we will also add a grpc backed `clientv3` function and a new `etcdctl` command.
+
+The metrics could look like:
+* `etcd_feature_enabled{name="featureName",stage="Alpha"} 1` if the feature is enabled.
+
+For grpc, new rpc type would be added to the `Maintenance` service.
+
+```proto
+service Maintenance {
+  ...
+  rpc FeatureGateStatus(FeatureGateStatusRequest) returns (FeatureGateStatusResponse) {
+    option (google.api.http) = {
+      post: "/v3/maintenance/featuregate"
+      body: "*"
+    };
+  }
+  ...
+}
+message Feature {
+  string name = 1;
+  bool enabled = 2;
+}
+message FeatureGateStatusRequest {
+  repeated string features = 2;
+}
+message FeatureGateStatusResponse {
+  ResponseHeader header = 1;
+  repeated Feature features = 3;
+}
+```
+
+The clientv3 function would look like:
+
+```go
+package clientv3
+
+type Maintenance interface {
+  FeatureGateStatus(ctx context.Context, features ...string) (*FeatureGateStatusResponse, error)
+}
+```
+
+The `etcdctl` commands could look like:
+* `etcdctl endpoint featuregate $featureName` returns true if the feature is enabled
+
+### Feature Gate
+
+We will use the new `k8s.io/component-base/featuregate.VersionedSpecs`(introduced in [kep-4330](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/4330-compatibility-versions)) to register and track the features along with their lifecycle at different release versions. 
```

Comment:
I think we should avoid a circular dependency on K8s.

Reply:
cc @dims

Reply:
agree, let's figure out a way *without* circular dependency!

Reply:
ok, we will copy the code to `pkg\`

Reply:
Sounds good, we can decide on long term solution when it's appropriate.